agent/cache/cache.go

```diff
@@ -515,9 +516,13 @@ func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ign
 		// code below. See https://github.com/hashicorp/consul/issues/4480.
 		newEntry.Error = err
 
-		if result.Value != nil {
+		labels := []metrics.Label{{Name: "updated", Value: strconv.FormatBool(result.EmptyCacheResult)}}
+		if result.Value != nil || result.EmptyCacheResult {
 			// A new value was given, so we create a brand new entry.
-			newEntry.Value = result.Value
+			if !result.EmptyCacheResult {
+				newEntry.Value = result.Value
+			}
```

Comment:
I starting look at how we might use this for the cache. 

`result.Value` is the full response (including `QueryMeta`), so it will never be nil if we have a cached result. The only way for it to be nil is an error.

For blocking queries, we never repeat the query unless the index has changed, right? So does this `AllowNotModifiedResponse` benefit us here in the cache? My understanding is that the cache would never make another request with the same index, so the logic that would send a `NotModified` response would never be invoked by a cache-type that supports blocking.

There are 2 that are non-blocking, so maybe we would benefit for those two?

Reply:
> result.Value is the full response (including QueryMeta), so it will never be nil if we have a cached result. The only way for it to be nil is an error.

Yes, I was probably too defensive here

> For blocking queries, we never repeat the query unless the index has changed, right?

Nope at all, at the maximum, every 10min, on a non-changing entry, you'll get the result. Assuming a watcher watching a service that do no change for 1h, 6 serializations in 1h. Not a big deal for 1 watcher, but multiply with the number of watchers (especially when Consul-Connect is enable and many agents are all watching the same services, this is a huge amount.

This might applies to non-blocking queries (eg: prepared queries)

Reply:
Ah, that makes sense. Would allowing the `RefreshTimeout` to be configurable be useful?

Reply:
@dnephin Well, having 10min is, for now, the max for a blocking query. Having less make for me little sense as we want to limit as much as we can (I made recently a PR in Prometheus to push this value at 10 min, see https://github.com/prometheus/prometheus/pull/7423 ) while having very good latency (as blocking queries return almost immediately when data is updated, so there is no added latency). Maybe on the long term it might be interresting to increase this value, but this probably goes beyond this PR.

Reply:
Makes sense. I'd like to have a look at that PR when I get a chance.

I've opened #8245 to use this in one cache-type